src/server/hotkey_collector.cpp

```diff
@@ -251,5 +261,111 @@ void hotkey_coarse_data_collector::analyse_data(detect_hotkey_result &result)
     }
 }
 
+hotkey_fine_data_collector::hotkey_fine_data_collector(
+    replica_base *base,
+    dsn::replication::hotkey_type::type hotkey_type,
+    int target_bucket_index,
+    int max_queue_size)
+    : internal_collector_base(base),
+      _hotkey_type(hotkey_type),
+      _max_queue_size(max_queue_size),
+      _target_bucket_index(target_bucket_index)
+{
+    // Distinguish between single-threaded and multi-threaded environments
+    if (_hotkey_type == dsn::replication::hotkey_type::READ) {
+
+        auto threads = dsn::get_threadpool_threads_info(THREAD_POOL_LOCAL_APP);
+        int queue_num = threads.size();
+
+        _string_capture_queue_vec.reserve(queue_num);
+        for (int i = 0; i < queue_num; i++) {
+            _thread_queue_map.insert(std::make_pair(threads[i]->native_tid(), i));
+
+            // Create a vector of the ReaderWriterQueue whose size = _max_queue_size
+            _string_capture_queue_vec.emplace_back(_max_queue_size);
+        }
+
+    } else { // WRITE
+        _string_capture_queue_vec.emplace_back(_max_queue_size);
+    }
+}
+
+void hotkey_fine_data_collector::capture_data(const dsn::blob &hash_key, uint64_t weight)
+{
+    if (get_bucket_id(hash_key) != _target_bucket_index) {
+        return;
+    }
+    _string_capture_queue_vec[get_queue_index()].try_emplace(std::make_pair(hash_key, weight));
+}
+
+struct blob_hash
+{
+    std::size_t operator()(const dsn::blob &str) const
+    {
+        dsn::string_view cp(str);
+        return boost::hash_range(cp.begin(), cp.end());
+    }
+};
+struct blob_equal
+{
+    std::size_t operator()(const dsn::blob &lhs, const dsn::blob &rhs) const
+    {
+        return dsn::string_view(lhs) == dsn::string_view(rhs);
+    }
+};
```

Comment:
Better to move to class `dsn::blob`'s definition, then you can reuse it easily.

Reply:
```cpp
    // blob.h
    bool operator==(const blob &r) const
    {
        // not implemented
        assert(false);
        return false;
    }
```
The default method is deleted, I think in different scenarios `==` has different meanings

Reply:
How the `==` operator of a class can have different behaviors?